### PR TITLE
rgw: improve handling of custom attributes exceeding the limit on name length

### PR DESCRIPTION
**Pull request by rzarzynski**
Fixes: http://tracker.ceph.com/issues/17935.
Additionally reduces the number of dynallocs on metadata extraction.

Currently undergoes testing.

CC: @cbodley, @mattbenjamin, @pritha-srivastava.

### CODE REVIEW

**Comment by rzarzynski** [comment]
s3-tests nor Tempest haven't found a regression here. Removing the *DNM* tag.

File: src/rgw/rgw_op.h [hunk 1]
```diff
@@ -1675,30 +1675,51 @@ static inline void format_xattr(std::string &xattr)
...
+      if (attr_name.length() > cct->_conf->osd_max_attr_name_len) {
+        return -ENAMETOOLONG;
+      }
+
+      using map_pair_t = std::map<std::string, ceph::bufferlist>::value_type;
+      using map_insert_ret_t =
+        std::pair<std::map<std::string, ceph::bufferlist>::iterator, bool>;
+
+      map_pair_t new_attr_pair(std::move(attr_name), ceph::bufferlist());
+      map_insert_ret_t rval = attrs.insert(std::move(new_attr_pair));
```
Comment (cbodley) [inline src/rgw/rgw_op.h]: can drop the typedefs and just use `auto rval = attrs.emplace(std::move(attr_name), bufferlist());`

File: src/rgw/rgw_rest_swift.cc [hunk 2]
```diff
@@ -1491,7 +1491,10 @@ void RGWInfo_ObjStore_SWIFT::list_swift_data(Formatter& formatter,
 
   string ceph_version(CEPH_GIT_NICE_VER);
   formatter.dump_string("version", ceph_version);
-  formatter.dump_int("max_meta_name_length", 81);
+
+  const size_t meta_name_limit = g_conf->osd_max_attr_name_len
+    - strlen(RGW_ATTR_PREFIX);
+  formatter.dump_int("max_meta_name_length", meta_name_limit);
```
Comment (pritha-srivastava) [inline src/rgw/rgw_rest_swift.cc]: @rzarzynski : The code looks fine, but as far as I remember, I had tried a command to set the attribute manually, to find the max length allowed, and I had found that it was 81. I will check that again and let you know?

File: src/rgw/rgw_rest_swift.cc [hunk 3]
```diff
@@ -1491,7 +1491,10 @@ void RGWInfo_ObjStore_SWIFT::list_swift_data(Formatter& formatter,
 
   string ceph_version(CEPH_GIT_NICE_VER);
   formatter.dump_string("version", ceph_version);
-  formatter.dump_int("max_meta_name_length", 81);
+
+  const size_t meta_name_limit = g_conf->osd_max_attr_name_len
+    - strlen(RGW_ATTR_PREFIX);
```
Comment (rzarzynski) [inline src/rgw/rgw_rest_swift.cc]: FIXME: should be `RGW_AMZ_META_PREFIX`. `RGW_ATTR_PREFIX` is only a substring of it.

File: src/rgw/rgw_rest_swift.cc (hunk 2 above)
Comment (rzarzynski) [inline src/rgw/rgw_rest_swift.cc]: @pritha-srivastava: on my environment the max length allowed is exactly 80 bytes (tested on RadosGW without the commits applied):
```
rgw$ curl -i "$publicURL/cont/" -X POST -H "X-Auth-Token: $token" -H "X-Container-Meta-`python -c 'print("a" * 81)'`: OK"
HTTP/1.1 500 Internal Server Error
Content-Length: 12
X-Trans-Id: tx00000000000000000000f-00583446c8-1016-default
Accept-Ranges: bytes
Content-Type: text/plain; charset=utf-8
Date: Tue, 22 Nov 2016 13:23:20 GMT

UnknownError

rgw$ curl -i "$publicURL/cont/" -X POST -H "X-Auth-Token: $token" -H "X-Container-Meta-`python -c 'print("a" * 80)'`: OK"
HTTP/1.1 204 No Content
X-Trans-Id: tx000000000000000000010-0058344ad0-1016-default
Content-Type: text/plain; charset=utf-8
Date: Tue, 22 Nov 2016 13:40:32 GMT

```

Output from `ngrep`:
```
T 127.0.0.1:54969 -> 127.0.0.1:8000 [AP]
POST /v1/KEY_2987047fc47840058e89e6182e0d96c3/cont/ HTTP/1.1.
User-Agent: curl/7.35.0.
Host: 127.0.0.1:8000.
Accept: */*.
X-Auth-Token: ee4025e5ca334747a0eaaa22e9b45e94.
X-Container-Meta-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa: OK.
.

##
T 127.0.0.1:8000 -> 127.0.0.1:54969 [AP]
HTTP/1.1 500 Internal Server Error.
Content-Length: 12.
X-Trans-Id: tx00000000000000000000f-00583446c8-1016-default.
Accept-Ranges: bytes.
Content-Type: text/plain; charset=utf-8.
Date: Tue, 22 Nov 2016 13:23:20 GMT.
.

##
T 127.0.0.1:8000 -> 127.0.0.1:54969 [AP]
UnknownError

########
T 127.0.0.1:54971 -> 127.0.0.1:8000 [AP]
POST /v1/KEY_2987047fc47840058e89e6182e0d96c3/cont/ HTTP/1.1.
User-Agent: curl/7.35.0.
Host: 127.0.0.1:8000.
Accept: */*.
X-Auth-Token: ee4025e5ca334747a0eaaa22e9b45e94.
X-Container-Meta-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa: OK.
.

##
T 127.0.0.1:8000 -> 127.0.0.1:54971 [AP]
HTTP/1.1 204 No Content.
X-Trans-Id: tx000000000000000000010-0058344ad0-1016-default.
Content-Type: text/plain; charset=utf-8.
Date: Tue, 22 Nov 2016 13:40:32 GMT.
.

```
Dump from the messenger:
```
2016-11-22 14:50:00.568233 7fc408817700  1 -- 127.0.0.1:0/3331651425 --> 127.0.0.1:6800/23074 -- osd_op(unknown.0.0:731 3.b843b263 .bucket.meta.298704
7fc47840058e89e6182e0d96c3:cont:89de0a15-d949-4289-99af-b81c2ac46155.4118.1 [delete,create 0~0,call version.check_conds,call version.set,writefull 0~4
35,setxattr user.rgw.acl (379),setxattr user.rgw.x-amz-meta-aaaaaaaaaaaaaaaaaaaaaa (3),setxattr user.rgw.x-amz-meta-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa (3),setxattr user.rgw.x-amz-meta-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaa (3)] snapc 0=[] ondisk+write+known_if_redirected e24) v7 -- 0x7fc4462fbe40 con 0
```

File: src/rgw/rgw_rest_swift.cc (hunk 3 above)
Comment (pritha-srivastava) [inline src/rgw/rgw_rest_swift.cc]: @rzarzynski : Lets say that the max allowed bytes is 80 (as tested by you), but the calculation here g_conf->osd_max_attr_name_len - strlen(RGW_AMZ_META_PREFIX) != 80 right? Please correct me if I am missing something here.

Comment (pritha-srivastava) [inline src/rgw/rgw_rest_swift.cc]: @rzarzynski : sounds fine now. Thanks!

File: src/rgw/rgw_op.h [hunk 4]
```diff
@@ -1675,30 +1675,46 @@ static inline void format_xattr(std::string &xattr)
...
-  map<string, string>::iterator iter;
-  for (iter = info.x_meta_map.begin(); iter != info.x_meta_map.end(); ++iter) {
-    const string &name(iter->first);
-    string &xattr(iter->second);
+  for (auto& kv : info.x_meta_map) {
+    const std::string& name = kv.first;
+    std::string& xattr = kv.second;
 
-    if (allow_empty_attrs || !xattr.empty()) {
+    if (allow_empty_attrs || ! xattr.empty()) {
```
Comment (oritwas) [inline src/rgw/rgw_op.h]: nit: space added

File: src/rgw/rgw_op.h [hunk 5]
```diff
@@ -1675,30 +1675,46 @@ static inline void format_xattr(std::string &xattr)
...
       lsubdout(cct, rgw, 10) << "x>> " << name << ":" << xattr << dendl;
+
```
Comment (oritwas) [inline src/rgw/rgw_op.h]: nit: extra line

File: src/rgw/rgw_rest_swift.cc [hunk 6]
```diff
@@ -1491,7 +1491,10 @@ void RGWInfo_ObjStore_SWIFT::list_swift_data(Formatter& formatter,
 
   string ceph_version(CEPH_GIT_NICE_VER);
   formatter.dump_string("version", ceph_version);
-  formatter.dump_int("max_meta_name_length", 81);
+
+  const size_t meta_name_limit = g_conf->osd_max_attr_name_len
+    - strlen(RGW_ATTR_PREFIX RGW_AMZ_META_PREFIX);
+  formatter.dump_int("max_meta_name_length", meta_name_limit);
```
Comment (oritwas) [inline src/rgw/rgw_rest_swift.cc]: good catch

File: src/rgw/rgw_op.h (hunk 4 above)
Comment (rzarzynski) [inline src/rgw/rgw_op.h]: FIXED.

File: src/rgw/rgw_op.h (hunk 5 above)
Comment (rzarzynski) [inline src/rgw/rgw_op.h]: FIXED.

**Comment by rzarzynski** [comment]
Closing in favour of #15369.

**Pull request closed without merging**